### PR TITLE
feat: pdp, category and serach events now extend pageevent

### PR DESCRIPTION
**Pull request by znikola**
The PDP, category and search events now extend the base `PageEvent`.

### CODE REVIEW

File: projects/storefrontlib/src/events/product/product-page-event.builder.ts [hunk 1]
```diff
@@ -72,22 +72,16 @@ export class ProductPageEventBuilder {
...
+        if (pageEvent.semanticRoute !== 'category') {
           return EMPTY;
         }
 
         return searchResults$.pipe(
           map((searchResults) => ({
-            categoryCode: pageEvent.categoryCode,
+            ...pageEvent,
+            categoryCode: pageEvent.context.id,
             categoryName: searchResults.breadcrumbs[0].facetValueName,
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/events/product/product-page-event.builder.ts]: can we drop the potential null pointer here? I've already had reports on that, not necessarily in this code.

File: projects/storefrontlib/src/events/product/product-page-event.builder.ts [hunk 2]
```diff
@@ -101,20 +95,24 @@ export class ProductPageEventBuilder {
...
+          return EMPTY;
+        }
 
-    return searchPageEvent$.pipe(
-      switchMap((isSearchPage) => (isSearchPage ? searchResults$ : EMPTY))
+        return searchResults$.pipe(
+          map((searchResults) => ({
+            ...pageEvent,
+            searchTerm: searchResults.freeTextSearch,
+            numberOfResults: searchResults.pagination.totalResults,
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/events/product/product-page-event.builder.ts]: I'd like to see the `numberOfResults` in the category page results as well. Not sure if that would result in more inheritance of the model.

**Review by tobi-or-not-tobi — DISMISSED** [review]
small things

File: projects/storefrontlib/src/events/product/product-page-event.builder.ts (hunk 1 above)
Comment (znikola) [reply projects/storefrontlib/src/events/product/product-page-event.builder.ts]: Good catch, done

File: projects/storefrontlib/src/events/product/product-page-event.builder.ts (hunk 2 above)
Comment (znikola) [reply projects/storefrontlib/src/events/product/product-page-event.builder.ts]: Sure, done.

File: projects/storefrontlib/src/events/product/product-page-event.builder.ts [hunk 3]
```diff
@@ -72,23 +72,20 @@ export class ProductPageEventBuilder {
...
+        if (pageEvent.semanticRoute !== 'category') {
           return EMPTY;
         }
 
         return searchResults$.pipe(
           map((searchResults) => ({
-            categoryCode: pageEvent.categoryCode,
-            categoryName: searchResults.breadcrumbs[0].facetValueName,
+            ...pageEvent,
+            categoryCode: pageEvent.context.id,
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/events/product/product-page-event.builder.ts]: what about:

```typescript
...{
  categoryCode: pageEvent?.context?.id,
  categoryName: searchResults.breadcrumbs?.[0].facetValueName,
  numberOfResults: searchResults.pagination?.totalResults,
}
```

as other values might be not available based on the occ param config.

File: projects/storefrontlib/src/events/product/product-page-event.builder.ts [hunk 4]
```diff
@@ -101,20 +98,24 @@ export class ProductPageEventBuilder {
...
+        if (pageEvent.semanticRoute !== 'search') {
+          return EMPTY;
+        }
 
-    return searchPageEvent$.pipe(
-      switchMap((isSearchPage) => (isSearchPage ? searchResults$ : EMPTY))
+        return searchResults$.pipe(
+          map((searchResults) => ({
+            ...pageEvent,
+            searchTerm: searchResults.freeTextSearch,
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/events/product/product-page-event.builder.ts]: same here i guess

**Review by tobi-or-not-tobi — CHANGES_REQUESTED** [review]

File: projects/storefrontlib/src/events/product/product-page-event.builder.ts [hunk 5]
```diff
@@ -72,23 +72,20 @@ export class ProductPageEventBuilder {
...
-        isCategoryPage: pageEvent.semanticRoute === 'category',
-        categoryCode: pageEvent.context.id,
-      }))
-    );
-
-    return categoryPageEvent$.pipe(
+    return this.eventService.get(PageEvent).pipe(
       switchMap((pageEvent) => {
-        if (!pageEvent.isCategoryPage) {
+        if (pageEvent?.semanticRoute !== 'category') {
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/events/product/product-page-event.builder.ts:77]: any reason to not filter?

File: projects/storefrontlib/src/events/product/product-page-event.builder.ts [hunk 6]
```diff
@@ -72,23 +72,20 @@ export class ProductPageEventBuilder {
...
           return EMPTY;
         }
 
         return searchResults$.pipe(
           map((searchResults) => ({
-            categoryCode: pageEvent.categoryCode,
-            categoryName: searchResults.breadcrumbs[0].facetValueName,
+            ...pageEvent,
+            categoryCode: pageEvent?.context?.id,
```
Comment (tobi-or-not-tobi) [inline projects/storefrontlib/src/events/product/product-page-event.builder.ts]: I means to add the `categoryCode` and `numberOfResults` to be conditionally added, similar to the `categoryName`

File: projects/storefrontlib/src/events/product/product-page-event.builder.ts (hunk 3 above)
Comment (znikola) [reply projects/storefrontlib/src/events/product/product-page-event.builder.ts]: Done

File: projects/storefrontlib/src/events/product/product-page-event.builder.ts (hunk 4 above)
Comment (znikola) [reply projects/storefrontlib/src/events/product/product-page-event.builder.ts]: Done

File: projects/storefrontlib/src/events/product/product-page-event.builder.ts (hunk 6 above)
Comment (znikola) [reply projects/storefrontlib/src/events/product/product-page-event.builder.ts]: Addressed above

File: projects/storefrontlib/src/events/product/product-page-event.builder.ts (hunk 5 above)
Comment (znikola) [reply projects/storefrontlib/src/events/product/product-page-event.builder.ts:77]: Yes, it breaks the stream. The idea is to _always_ switch from the base `PageEvent` and then further inspect the event and conditionally return an appropriate stream (`EMPTY` or the actual event stream)

**Review by tobi-or-not-tobi — APPROVED** [review]

**Pull request merged**